File: common/src/main/java/net/shiroha233/roadweaver/features/path/pathlogic/pathfinding/TerrainSamplingCache.java
package net.shiroha233.roadweaver.features.path.pathlogic.pathfinding;

import net.minecraft.core.Holder;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.tags.BiomeTags;
import net.minecraft.world.level.biome.Biome;

import java.util.concurrent.ConcurrentHashMap;

/**
 * 地形采样缓存 - 使用 FastHeightSampler 进行高性能采样
 * 
 * 性能优化：
 * - 使用 FastHeightSampler 替代原版 getBaseHeight()，提速 10-50x
 * - 原版每次调用都创建新的 NoiseChunk（极其昂贵）
 * - 新方案直接使用 DensityFunction 采样，绕过 NoiseChunk 创建开销
 * 
 * 线程安全：
 * - 使用 ConcurrentHashMap 保证多线程访问安全
 */
public final class TerrainSamplingCache {
    private final ConcurrentHashMap<Long, Boolean> waterCache = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<Long, Boolean> nearWaterCache = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<Long, Boolean> columnWaterCache = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<Long, Integer> heightCache = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<Long, Integer> oceanFloorCache = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<Long, Holder<Biome>> biomeCache = new ConcurrentHashMap<>();
    
    // 快速高度采样器（惰性初始化，volatile 保证可见性）
    private volatile FastHeightSampler fastSampler;

    private static long hashXZ(int x, int z) {
        return ((long) x << 32) | (z & 0xffffffffL);
    }

    /**
     * 获取地表高度 - 使用快速采样
     */
    public int height(ServerLevel level, int x, int z) {
        long key = hashXZ(x, z);
        Integer cached = heightCache.get(key);
        if (cached != null) {
            TerrainSamplingStats.recordCacheHit();
            return cached;
        }
        TerrainSamplingStats.recordCacheMiss();
        
        // 惰性初始化快速采样器（双重检查锁定）
        if (fastSampler == null) {
            synchronized (this) {
                if (fastSampler == null) {
                    fastSampler = FastHeightSampler.create(level);
                }
            }
        }
        
        int h = fastSampler.sampleHeight(x, z);
        heightCache.put(key, h);
        return h;
    }

    public boolean isWaterLike(ServerLevel level, int x, int z) {
        long key = hashXZ(x, z);
        Boolean cached = waterCache.get(key);
        if (cached != null) {
            TerrainSamplingStats.recordCacheHit();
            return cached;
        }
        TerrainSamplingStats.recordCacheMiss();

        var chunkSource = level.getChunkSource();
        var randomState = chunkSource.getGeneratorState().randomState();
        var biomeSource = chunkSource.getGenerator().getBiomeSource();
        Holder<Biome> biome = biomeSource.getNoiseBiome(x >> 2, 16, z >> 2, randomState.sampler());

        boolean res = biome.is(BiomeTags.IS_RIVER) || biome.is(BiomeTags.IS_OCEAN) || biome.is(BiomeTags.IS_DEEP_OCEAN);
        waterCache.put(key, res);
        return res;
    }

    /**
     * 获取海底/河床高度 - 使用快速采样
     */
    public int oceanFloor(ServerLevel level, int x, int z) {
        long key = hashXZ(x, z);
        Integer cached = oceanFloorCache.get(key);
        if (cached != null) {
            TerrainSamplingStats.recordCacheHit();
            return cached;
        }
        TerrainSamplingStats.recordCacheMiss();
        
        // 惰性初始化快速采样器（双重检查锁定）
        if (fastSampler == null) {
            synchronized (this) {
                if (fastSampler == null) {
                    fastSampler = FastHeightSampler.create(level);
                }
            }
        }
        
        // FastHeightSampler 使用 initialDensityWithoutJaggedness，
        // 不区分 OCEAN_FLOOR 和 MOTION_BLOCKING，对道路寻路足够
        int h = fastSampler.sampleHeight(x, z);
        oceanFloorCache.put(key, h);
        return h;
    }

    public boolean isNearWaterLike(ServerLevel level, int x, int z, int neighborDistance) {
        long key = hashXZ(x, z);
        Boolean cached = nearWaterCache.get(key);
        if (cached != null) {
            TerrainSamplingStats.recordCacheHit();
            return cached;
        }
        TerrainSamplingStats.recordCacheMiss();
        int d = neighborDistance;
        int[][] neighborOffsets = new int[][] {
                { d, 0 }, { -d, 0 }, { 0, d }, { 0, -d },
                { d, d }, { d, -d }, { -d, d }, { -d, -d }
        };
        for (int[] off : neighborOffsets) {
            int nx = x + off[0];
            int nz = z + off[1];
            if (isWaterLike(level, nx, nz)) {
                nearWaterCache.put(key, true);
                return true;
            }
        }
        nearWaterCache.put(key, false);
        return false;
    }

    public boolean isColumnWater(ServerLevel level, int x, int z) {
        long key = hashXZ(x, z);
        Boolean cached = columnWaterCache.get(key);
        if (cached != null) {
            TerrainSamplingStats.recordCacheHit();
            return cached;
        }
        TerrainSamplingStats.recordCacheMiss();

        int of = oceanFloor(level, x, z);
        int h = height(level, x, z);
        int sea = level.getSeaLevel();

        // 方法1：群系判断
        boolean isWaterBiome = isWaterLike(level, x, z);
        boolean biomeWater = isWaterBiome && of < sea;

        // 方法2：高度差判断
        boolean heightWater = (h <= sea + 1) && (of < h - 1);

        boolean res = biomeWater || heightWater;
        columnWaterCache.put(key, res);
        return res;
    }

    public Holder<Biome> getBiome(ServerLevel level, int x, int z) {
        long key = hashXZ(x, z);
        Holder<Biome> cached = biomeCache.get(key);
        if (cached != null) {
            TerrainSamplingStats.recordCacheHit();
            return cached;
        }
        TerrainSamplingStats.recordCacheMiss();
        var chunkSource = level.getChunkSource();
        var randomState = chunkSource.getGeneratorState().randomState();
        var biomeSource = chunkSource.getGenerator().getBiomeSource();
        Holder<Biome> biome = biomeSource.getNoiseBiome(x >> 2, 16, z >> 2, randomState.sampler());
        biomeCache.put(key, biome);
        return biome;
    }

    /**
     * 预热指定区域的高度缓存
     */
    public void prewarmRegion(ServerLevel level, int minX, int minZ, int maxX, int maxZ, int step) {
        if (fastSampler == null) {
            synchronized (this) {
                if (fastSampler == null) {
                    fastSampler = FastHeightSampler.create(level);
                }
            }
        }
        fastSampler.prewarmRegion(minX, minZ, maxX, maxZ, step);
    }

    public void clear() {
        waterCache.clear();
        nearWaterCache.clear();
        columnWaterCache.clear();
        heightCache.clear();
        oceanFloorCache.clear();
        biomeCache.clear();
        
        if (fastSampler != null) {
            fastSampler.clearCache();
        }
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/features/path/pathlogic/pathfinding/TerrainSamplingCache.kt
package net.shiroha233.roadweaver.features.path.pathlogic.pathfinding

import net.minecraft.core.Holder
import net.minecraft.server.level.ServerLevel
import net.minecraft.tags.BiomeTags
import net.minecraft.world.level.biome.Biome
import net.minecraft.world.level.levelgen.Heightmap

class TerrainSamplingCache {
    private val waterCache = HashMap<Long, Boolean>()
    private val nearWaterCache = HashMap<Long, Boolean>()
    private val columnWaterCache = HashMap<Long, Boolean>()
    private val heightCache = HashMap<Long, Int>()
    private val oceanFloorCache = HashMap<Long, Int>()
    private val biomeCache = HashMap<Long, Holder<Biome>>()

    companion object {
        private fun hashXZ(x: Int, z: Int): Long {
            return (x.toLong() shl 32) or (z.toLong() and 0xffffffffL)
        }
    }

    fun height(level: ServerLevel, x: Int, z: Int): Int {
        val key = hashXZ(x, z)
        heightCache[key]?.let {
            TerrainSamplingStats.recordCacheHit()
            return it
        }
        TerrainSamplingStats.recordCacheMiss()
        val generator = level.chunkSource.generator
        val rs = level.chunkSource.generatorState.randomState()
        val sea = level.seaLevel
        val motion = generator.getBaseHeight(x, z, Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, level, rs)
        if (motion > sea + 2) {
            heightCache[key] = motion
            return motion
        }

        val surface = generator.getBaseHeight(x, z, Heightmap.Types.WORLD_SURFACE_WG, level, rs)
        val waterBiome = isWaterLike(level, x, z)
        val shouldUseSurface = waterBiome || (surface <= sea + 2 && oceanFloor(level, x, z) < sea)
        val h = if (shouldUseSurface) surface else motion

        heightCache[key] = h
        return h
    }

    fun isWaterLike(level: ServerLevel, x: Int, z: Int): Boolean {
        val key = hashXZ(x, z)
        waterCache[key]?.let {
            TerrainSamplingStats.recordCacheHit()
            return it
        }
        TerrainSamplingStats.recordCacheMiss()

        // 修正：使用 BiomeSource 进行噪声采样，不加载区块
        // 注意：getNoiseBiome 需要夸脱坐标 (x >> 2, y >> 2, z >> 2)
        val chunkSource = level.chunkSource
        val randomState = chunkSource.generatorState.randomState()
        val biomeSource = chunkSource.generator.biomeSource

        // 采样 Y=64 处的生物群系（海平面附近）
        val biome = biomeSource.getNoiseBiome(x shr 2, 16, z shr 2, randomState.sampler())

        val res = biome.`is`(BiomeTags.IS_RIVER) || biome.`is`(BiomeTags.IS_OCEAN) || biome.`is`(BiomeTags.IS_DEEP_OCEAN)
        waterCache[key] = res
        return res
    }

    fun oceanFloor(level: ServerLevel, x: Int, z: Int): Int {
        val key = hashXZ(x, z)
        oceanFloorCache[key]?.let {
            TerrainSamplingStats.recordCacheHit()
            return it
        }
        TerrainSamplingStats.recordCacheMiss()
        val generator = level.chunkSource.generator
        val rs = level.chunkSource.generatorState.randomState()
        // 修正：使用对应的 Heightmap 类型进行噪声采样
        val h = generator.getBaseHeight(x, z, Heightmap.Types.OCEAN_FLOOR_WG, level, rs)
        oceanFloorCache[key] = h
        return h
    }

    fun isNearWaterLike(level: ServerLevel, x: Int, z: Int, neighborDistance: Int): Boolean {
        val key = hashXZ(x, z)
        nearWaterCache[key]?.let {
            TerrainSamplingStats.recordCacheHit()
            return it
        }
        TerrainSamplingStats.recordCacheMiss()
        val d = neighborDistance
        val neighborOffsets = arrayOf(
            intArrayOf(d, 0), intArrayOf(-d, 0), intArrayOf(0, d), intArrayOf(0, -d),
            intArrayOf(d, d), intArrayOf(d, -d), intArrayOf(-d, d), intArrayOf(-d, -d)
        )
        for (off in neighborOffsets) {
            val nx = x + off[0]
            val nz = z + off[1]
            if (isWaterLike(level, nx, nz)) {
                nearWaterCache[key] = true
                return true
            }
        }
        nearWaterCache[key] = false
        return false
    }

    fun isColumnWater(level: ServerLevel, x: Int, z: Int): Boolean {
        val key = hashXZ(x, z)
        columnWaterCache[key]?.let {
            TerrainSamplingStats.recordCacheHit()
            return it
        }
        TerrainSamplingStats.recordCacheMiss()

        // 使用多种方式检测水体，解决以下问题：
        // 1. 浅滩(beach)不在 IS_RIVER/IS_OCEAN 群系标签中，但实际有水
        // 2. 群系边界处噪声采样可能判断失误，导致跨海不建桥

        val of = oceanFloor(level, x, z)  // OCEAN_FLOOR_WG：海底/河床高度
        val h = height(level, x, z)        // MOTION_BLOCKING_NO_LEAVES：表面高度
        val sea = level.seaLevel

        // 方法1：群系判断（原有逻辑）
        // 适用于标准的河流/海洋群系
        val isWaterBiome = isWaterLike(level, x, z)
        val biomeWater = isWaterBiome && of < sea

        // 方法2：高度差判断（新增逻辑）
        // 核心原理：如果表面高度接近海平面，但海底明显更低，说明中间是水
        // - h <= sea + 1：表面在海平面或略高（水面通常在 seaLevel）
        // - of < h - 1：海底比表面低至少2格，说明有水深
        // 这样可以检测到：浅滩、沼泽边缘、甚至非标准群系中的水体
        val heightWater = (h <= sea + 1) && (of < h - 1)

        // 任一方法检测到水体即可
        val res = biomeWater || heightWater

        columnWaterCache[key] = res
        return res
    }

    fun getBiome(level: ServerLevel, x: Int, z: Int): Holder<Biome> {
        val key = hashXZ(x, z)
        biomeCache[key]?.let {
            TerrainSamplingStats.recordCacheHit()
            return it
        }
        TerrainSamplingStats.recordCacheMiss()
        val chunkSource = level.chunkSource
        val randomState = chunkSource.generatorState.randomState()
        val biomeSource = chunkSource.generator.biomeSource
        val biome = biomeSource.getNoiseBiome(x shr 2, 16, z shr 2, randomState.sampler())
        biomeCache[key] = biome
        return biome
    }

    fun clear() {
        waterCache.clear()
        nearWaterCache.clear()
        columnWaterCache.clear()
        heightCache.clear()
        oceanFloorCache.clear()
        biomeCache.clear()
    }
}
